### ISSUE TITLE
Fails to commit with signing

### ISSUE THREAD

**Issue by Mange** [TimUntersberger/neogit#370]
### Description

When committing using Neogit (<kbd>c</kbd><kbd>c</kbd>) the commit fails when signing is enabled.

**Still works:**

1. Disabling commit signing with <kbd>-</kbd><kbd>S</kbd> first.
2.  Committing inside Neovim with a `!` command: `:!git commit --allow-empty -m "Test"`.
3. Placing Neovim in the background with <kbd>Ctrl</kbd>+<kbd>Z</kbd> and committing from the shell.

Relevant Git config:

```ini
[user]
	signingkey = 2EA6F4AA110A1BF2227519A90443C69F6F022CDE

[commit]
	gpgsign = true
```

When committing in any working fashion, I'm not prompted for a password or anything as that is handled through my GPG agent. I somewhat suspect this might be the point where it fails, though. When Neogit is shelling out, is it maybe not using the same environment or shell as `:set shell`?

This used to work. It broke last week, or thereabouts, but I also don't bump Neovim plugins very often so the change can be older. Maximum of around 2 months.

How can I help in debugging this?

### Neovim version

```
NVIM v0.9.0-dev-72-gc5347d47e1
Build type: RelWithDebInfo
LuaJIT 2.1.0-beta3
Compilation: /usr/bin/cc -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fexceptions     -Wp,-D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security     -fstack-clash-protection -fcf-protection -Wp,-U_FORTIFY_SOURCE -Wp,-D_FORTIFY_SOURCE=1 -DNVIM_TS_HAS_SET_MATCH_LIMIT -DNVIM_TS_HAS_SET_ALLOCATOR -O2 -g -Og -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wdouble-promotion -Wmissing-noreturn -Wmissing-format-attribute -Wmissing-prototypes -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fno-common -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DMIN_LOG_LEVEL=3 -I/home/mange/.cache/paru/clone/neovim-git/src/build/cmake.config -I/home/mange/.cache/paru/clone/neovim-git/src/neovim-git/src -I/usr/include -I/home/mange/.cache/paru/clone/neovim-git/src/build/src/nvim/auto -I/home/mange/.cache/paru/clone/neovim-git/src/build/include
```

**NOTE:** This also happened on a nightly version pre-0.8 stable. Upgrading to latest stable was my first try to get this working.

### Operating system and version

```
Linux socia 5.19.6-arch1-1 #1 SMP PREEMPT_DYNAMIC Wed, 31 Aug 2022 22:09:40 +0000 x86_64 GNU/Linux
```

### Steps to reproduce

Make sure you have commit signing set up first. You should be able to sign using your local shell.

1. `mkdir /tmp/git-test`
2. `cd /tmp/git-test`
3. `echo First > test.txt`
4. `git init && git add . && git commit -m "Initial commit"`
5. Verify that the commit was signed. `git show`
6. `nvim -nu /path/to/minimal.lua test.txt`
7. Make a change.
8. `:Neogit`
9. <kbd>c</kbd><kbd>c</kbd>
10. Write a message, and commit by saving and closing buffer.

### Expected behavior

Commit is created and signed.

### Actual behavior

Commit fails.

**Command history after failed commit**

```
  0 git --no-optional-locks ls-files --others --cached --modified --deleted --full-name .git/COMMIT_EDITMSG (0.004 ms) [stdout 0]
128 git --no-optional-locks commit (1.399 ms) [stderr 2]
  | error: gpg failed to sign the data
  | fatal: failed to write commit object
  0 git --no-optional-locks status --porcelain=2 -b -z (0.001 ms) [stdout 1]
```

Written commit message is lost if one tries to commit again, making this really annoying.

### Minimal config

```Lua
vim.cmd([[set runtimepath=$VIMRUNTIME]])
vim.cmd([[set packpath=/tmp/nvim/site]])
local package_root = "/tmp/nvim/site/pack"
local install_path = package_root .. "/packer/start/packer.nvim"
local function load_plugins()
  require("packer").startup({
    {
      "wbthomason/packer.nvim",
      {
        "TimUntersberger/neogit",
        requires = {
          { "nvim-lua/plenary.nvim" },
        },
        config = function()
          print("loaded neogit")
          require("neogit").setup()
        end,
      },
    },
    config = {
      package_root = package_root,
      compile_path = install_path .. "/plugin/packer_compiled.lua",
      display = { non_interactive = true },
    },
  })
end
if vim.fn.isdirectory(install_path) == 0 then
  print("Installing neogit and dependencies.")
  vim.fn.system({ "git", "clone", "--depth=1", "https://github.com/wbthomason/packer.nvim", install_path })
end
load_plugins()
require("packer").sync()
```

**Comment by Mange** [comment]
This seems to have stopped happning for me now. I think it might be because I restarted the computer, but I'm not sure if that's the only thing that happened.

I'll close for now, but reopen if I manage to figure out what caused it.

**Comment by itmecho** [comment]
I'm getting this issue and I think it's because the required environment variables for gpg to prompt for a password are no longer passed to the commit command. If I commit manually in the terminal, enter my password, then commit via neogit afterwards, it works as the password is cached in the gpg-agent.

**Comment by Mange** [comment]
I think it might be weirder than that. It might be because my window manager had restarted in a way that may have lost some environment variables that got re-populated when I opened new terminals.

Then that it looked like `^Z` in my editor terminal working might have been me not realizing that the agent was unlocked at that point.

I don't have password prompts anywhere, it's unlocked automatically by my wallet process. So it's not as simple as it not reading stdin; nothing is prompted. It just never asked the wallet properly, I think.

**Comment by Mange** [comment]
Okay, so I've tested this some more now.

For sure, the committing through Neogit directly breaks when the encryption keys are locked again, and fails to unlock. However, just doing `!git commit -m "test"` still works without prompting.

Right now ENV is reset and only a small subset is passed along:
https://github.com/xanderio/neogit/blob/463820a83f4ba387655f370a17c87dc3100cdf0d/lua/neogit/process.lua#L31-L37

I suggest we also pass along important variables like `XAUTHORITY` and `DISPLAY` (so password prompts and wallets work) and `SSH_AUTH_SOCK` so SSH agents also work.

**Comment by ten3roberts** [comment]
It seems that you are using a fork of neogit.

All envs are now passed similar to `:!` as a few days ago just because of gpg issues and precommit requiring a modified path, such as python, go, or nixos

**Comment by Mange** [comment]
Oh, yeah. I accidentally linked to the permalink of a forked repo there. I was too quick with building the link for the comment. :upside_down_face: 

Seems to be working now that I pulled the latest version. Thank you for letting me know! :heart: 

(I was on the latest version as of the opening of this issue. The problem was fixed somewhere between `74c9e29..1acb13c` if anyone else also wants to figure it out.)
